lib/FrontendTool/FrontendTool.cpp

```diff
@@ -623,6 +623,116 @@ static void verifyGenericSignatures(CompilerInvocation &Invocation,
   }
 }
 
+static void dumpOneScopeMapLocation(unsigned bufferID,
+                                    std::pair<unsigned, unsigned> lineColumn,
+                                    SourceManager &sourceMgr, ASTScope &scope) {
+  SourceLoc loc =
+      sourceMgr.getLocForLineCol(bufferID, lineColumn.first, lineColumn.second);
+  if (loc.isInvalid())
+    return;
+
+  llvm::errs() << "***Scope at " << lineColumn.first << ":" << lineColumn.second
+               << "***\n";
+  auto locScope = scope.findInnermostEnclosingScope(loc);
+  locScope->print(llvm::errs(), 0, false, false);
+
+  // Dump the AST context, too.
+  if (auto dc = locScope->getDeclContext()) {
+    dc->printContext(llvm::errs());
+  }
+
+  // Grab the local bindings introduced by this scope.
+  auto localBindings = locScope->getLocalBindings();
+  if (!localBindings.empty()) {
+    llvm::errs() << "Local bindings: ";
+    interleave(localBindings.begin(), localBindings.end(),
+               [&](ValueDecl *value) { llvm::errs() << value->getFullName(); },
+               [&]() { llvm::errs() << " "; });
+    llvm::errs() << "\n";
+  }
+}
+
+static void dumpAndPrintScopeMap(CompilerInvocation &Invocation,
+                                 CompilerInstance &Instance, SourceFile *SF) {
+  ASTScope &scope = SF->getScope();
+
+  if (Invocation.getFrontendOptions().DumpScopeMapLocations.empty()) {
+    scope.expandAll();
+  } else if (auto bufferID = SF->getBufferID()) {
+    SourceManager &sourceMgr = Instance.getSourceMgr();
+    // Probe each of the locations, and dump what we find.
+    for (auto lineColumn :
+         Invocation.getFrontendOptions().DumpScopeMapLocations)
+      dumpOneScopeMapLocation(*bufferID, lineColumn, sourceMgr, scope);
+
+    llvm::errs() << "***Complete scope map***\n";
+  }
+  // Print the resulting map.
+  scope.print(llvm::errs());
+}
+
+static SourceFile *getPrimaryOrMainSourceFile(CompilerInvocation &Invocation,
+                                              CompilerInstance &Instance) {
+  SourceFile *SF = Instance.getPrimarySourceFile();
+  if (!SF) {
+    SourceFileKind Kind = Invocation.getSourceFileKind();
+    SF = &Instance.getMainModule()->getMainSourceFile(Kind);
+  }
+  return SF;
+}
+
+/// We've been told to dump the AST (either after parsing or type-checking,
+/// which is already differentiated in CompilerInstance::performSema()),
+/// so dump or print the main source file and return.
+
+static Optional<bool> dumpAST(CompilerInvocation &Invocation,
```

Comment:
Can we can this `maybeDumpAST`, since it checks to see if the Action is one of the dump actions?

Reply:
There are other routines like this & I would like to be consistent. We could use the "maybe" prefix, or the "IfRequested" suffix, or the "IfNeeded" suffix. Which would be best?

Reply:
Ah, yeah, the "IfNeeded" you're using elsewhere is fine.